### ISSUE TITLE
Flatpak Jazz grap-bag of related issues

### ISSUE THREAD

**Issue by Eonfge** [FontManager/font-manager#122]
I started working on a Flathub package for Font Manager, and I noticed that the application id (or slug as some call it) is `org.gnome.FontViewer`. This is a bit weird though, since the application is not actually hosted on gnome.org and as far as I know it' not an official GNOME package.

Is this for backwards compatibility reasons?

Also, would it be possible to change the application id, and all that is related  to it, to `io.github.FontViewer`? This would be in compliance with the standard and it would make it easier to make a Flathub package.

https://www.freedesktop.org/software/appstream/docs/chap-Quickstart.html

**Comment by JerryCasiano** [comment]
Feel to free to file a pull request. I'll be happy to review.

That ID was adopted long before GNOME had any guidelines addressing this and taking into consideration that some core GNOME developers had stated on a mailing list at the time that an application like this was out of scope and should be done by a third party so I don't plan on doing any extra work to change it now. If someone else does so without causing any issues for users I'll be happy to merge the changes, otherwise it's not happening.

Also, I would very much appreciate and ask that you ensure that any package you decide to upload is completely functional and up to date. 

I would certainly not appreciate anyone uploading a broken Flatpak, Snap, or <insert whatever package format here> as this does more harm than good.

.
.Thanks.

**Comment by JerryCasiano** [comment]
#73

**Comment by Eonfge** [comment]
Ow, that's good info. I was already looking in giving proper access to font-config and such. And fear not, I won't half-arse it and this is not my first application.

My work-in-progress branch is here:
https://github.com/Eonfge/flathub/tree/org.gnome.FontViewer

---

I can btw confirm the compiler-issue that is mentioned in the other thread. There is a problem with the linker and the unicode-library. Make a different issue for that?

**Comment by JerryCasiano** [comment]
> lI can btw confirm the compiler-issue that is mentioned in the other thread. There is a problem with the linker and the unicode-library. Make a different issue for that?

Are you building using the latest commit?

You can post any details / output you get here but I'm thinking I worked around this in master.

**Comment by Eonfge** [comment]
using the `master` and `ed259fec1a7b683214b33cbed5649bbe5219c074` works. so I guess that 0.7.8 would be good.

![Screenshot from 2020-06-20 22-55-43](https://user-images.githubusercontent.com/33983090/85211414-492ea680-b349-11ea-879a-3d4a11680682.png)

I'll start testing it tomorrow... or later, but compiling and running in a flatpak works. I did change the branch to: https://github.com/Eonfge/flathub/tree/org.gnome.FontManager so that the Manager is packaged and not the viewer... do we want both utilities? We could make two different Flatpaks from that.

**Comment by JerryCasiano** [comment]
Entirely up to you.

The latest commit also allows you to specify that you only want one or the other as far as manager and viewer go so that should make separate packaging fairly easy. 

I can do a release if you need me to for whatever reason but I'd rather not since there really aren't any major differences yet.

I feel like you could just use the latest commit as 0.7.7, I realize it may be frowned upon since it won't match the actual tag in the repo but like I said there aren't really any significant changes and since the commit is specified in the manifest I don't see it as a big deal.

By the way, if you do run into runtime issues, post about them here and we'll see if we can work around them somehow as long as it doesn't affect non-Flatpak builds I'm up for it. We can keep this open till it's published.

Thanks!

**Comment by JerryCasiano** [comment]
Also there are screenshots at https://github.com/FontManager/resources

Not sure if you need any or they even are the right format but figured I'd mention it.

**Comment by Eonfge** [comment]
Thanks for linking the images. By default, Flathub uses the images from the meta.xml.

On the topic of the meta.xml, I made a pull request to add empty age ratings. This makes it clear that the application has no age restrictions. It's a required component for Flathub

**Comment by JerryCasiano** [comment]
> On the topic of the meta.xml, I made a pull request to add empty age ratings. This makes it clear that the application has no age restrictions. It's a required component for Flathub

Done. Thanks.

**Comment by Eonfge** [comment]
![Screenshot from 2020-06-23 18-36-35](https://user-images.githubusercontent.com/33983090/85430609-a3657c80-b580-11ea-891e-377c27c0621b.png)

Making good progress: 
- You can configure all fonts on your machines, 
- you can make overrides, and
- you can import custom fonts from `~/Fonts`

Things I must still configure:
- GSettings access

**Comment by JerryCasiano** [comment]
Nice.

I didn't expect changes would persist without changes to the code.

Admittedly I didn't spend much time on it.

Thanks for the update.

**Comment by Eonfge** [comment]
@JerryCasiano I have found something that would be nice to implement, which should also not affect existing applications; File portals.

Right now, you read the folder directly from the location where the user indicated that it is, instead of reading it through a portal. The portal is (in case of GNOME) a special instance of the Nautilus file/folder finder that will give you the result of the user-input, without requiring you to directly read the given path.

Coincidentally, this is also fully back- and forward compatible with Flatpak. It will allow a user to open any directory containing fonts, and it will then allow them to also view a font with Font Manager directly (using the mimetype-matcher)

---

Currently, the things I found that won't work:
1. Right click 'show in folder'
2. Select source from outside font-related paths 
 _(I added .fonts for backwards compatibility)_
3. Mimetype 'open with Font Manager'

2 and 3 can be fixed using a file-portal, but that would require you to change the current application. Else, I don't think they are that critical and I wouldn't hold a Flathub package back for it, how do you look at this?

_Having some understanding of Vala myself, you'll have to use a GTK FileChooser with FILE_CHOOSER_ACTION_SELECT_FOLDER_

**Comment by JerryCasiano** [comment]
> Else, I don't think they are that critical and I wouldn't hold a Flathub package back for it, how do you look at this?

1 & 3 I don't consider critical, though the show in folder option not working does give a bad impression.

2 I think is kind of a big deal, I would definitely like that to work properly if possible.

If that can be resolved by using portals, that would be the way to go.

We use standard GTK file dialogs so I'm thinking this shouldn't be a big deal.

I'll look into it and see if we can get that done.

**Comment by JerryCasiano** [comment]
If you get a chance please test with the latest commit, which switches our file selection dialogs to GtkFileChooserNative,  and see if that helps the issue with Font Sources. I believe that's all the changes that are required on our end.

I'll build a Flatpak using your work within the next couple days and see if the "Show in Folder" issue is something that we can work around.

Thanks.

**Comment by JerryCasiano** [comment]
Well, that's interesting... 

According to https://docs.flatpak.org/en/latest/portals-gtk.html gtk_show_uri_on_window() is the recommended way to open URI's and that's exactly what we already do in our "Show in Folder" method.

Not sure why it's not actually working, I guess we'll have to switch to g_app_info_launch_default_for_uri(0 and see if that does the trick.

**Comment by Eonfge** [comment]
Works! In one fucking try! That's honestly quite rare :partying_face: 

![Screenshot from 2020-06-25 18-06-16](https://user-images.githubusercontent.com/33983090/85755872-9d050b00-b70e-11ea-92b6-51d64ec5d6d3.png)

This would fix issue 2 in one quick stroke.

**Comment by Eonfge** [comment]
_This is turning more-and-more into a complete project review, so please tell me if I'm becoming too demanding_

Have you btw thought about your application icon? Right now, you use a default `preferences-desktop-font` from Adawaita... but that's a deprecated standard. I can include the original image in the package, essentially just copy-pasting it from Adawaita, but you might want to make it more 'official' and just add it to the project yourself.

What do you think? 
- Keep it, problem of downstream
- Keep it, include in project
- Other...

**Comment by JerryCasiano** [comment]
> This is turning more-and-more into a complete project review, so please tell me if I'm becoming too demanding

Not at all. Anything that helps packaging efforts is almost always an improvement worth doing. 
Especially in this case, I would much rather use API that works universally than anything else, that's win-win.

> Have you btw thought about your application icon?

I have. 

I took the lazy way out and just decided to depend on the users icon theme because... well, it's easy, it also respects the users theme choice but mostly because we don't actually have an icon.

I'm open to just importing it into the project though. I'll look into what changes are needed later today or tomorrow.

**Comment by Eonfge** [comment]
The newer standard of icons links them directly to the application id. If you install an icon called `org.gnome.FontManager.png` into '/app/share/icons/hicolor/128x128/apps/', then that's completely compliant to the standard. 

You can actually see me do that in the .yaml file:
https://github.com/Eonfge/flathub/blob/71bd63eae91ebcf8ada2480c2603daadab6bd659/org.gnome.FontManager.yaml#L52

**Comment by JerryCasiano** [comment]
I think we should be good on the icon issue.

**Comment by Eonfge** [comment]
This is getting really smooth, your cooperation is really praiseworthy.

I was thinking, perhaps it's time to submit this to Flathub. Perhaps making a new release version for this is smart though, many other distributions will also want to have the changes to the native File Chooser and the icons for example.

Also, would you like to have co-maintainer status?

**Comment by JerryCasiano** [comment]
> I was thinking, perhaps it's time to submit this to Flathub. Perhaps making a new release version for this is smart though, many other distributions will also want to have the changes to the native File Chooser and the icons for example.

I'm on board with this.

But, there are a few changes that I would want to include in the next release so... give me two weeks on this.

> Also, would you like to have co-maintainer status?

I appreciate the offer but since I'm not familiar at all with Flatpak or Flathub I would prefer not to at this point.

**Comment by Eonfge** [comment]
Is good. I'll wait for the next release, before creating a Flathub submission.

That's fine, you're always welcome to get directly involved if you want.

**Comment by JerryCasiano** [comment]
So I wasn't able to finish up on the changes I was hoping to include but I did draft a release today anyway.

Please test using the latest release and feel free to close this if all is well.

Thanks again!

**Comment by Eonfge** [comment]
Looks all good to me. Time to submit it I would say :smiley:

**Comment by Eonfge** [comment]
The only thing in our way, is the missing release info

Would it be possible to add this, or should I try to fix it in post?

```
<releases>
  <release version="0.7.8" date="2020-07-10"/>
</releases>
```

You can of cause add more info to the [release-tag](https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-releases), but I'm not bothered by it being empty

**Comment by Eonfge** [comment]
I've gone ahead and I've fixed the release version in post. So, with not further delay, the submission:
https://github.com/flathub/flathub/pull/1646

**Comment by JerryCasiano** [comment]
You're right it validates successfully now, not sure why it didn't last time I tried it.

The <icon> tag still causes it to fail validation, so either I'm misunderstanding the spec, I'm doing something else wrong or validation is still buggy.

`• tag-invalid           : <icon> not allowed in desktop appdata`

https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-icon

**Comment by JerryCasiano** [comment]
By the way, is there a way to allow a flatpak access to dconf settings?

It would be nice for the Gnome desktop font settings to work.

**Comment by Eonfge** [comment]
For the icon, I would just leave it out. Icons are mostly part of the .desktop file. I've never seen the icon tag used in the metainfo.

> By the way, is there a way to allow a flatpak access to dconf settings?

:pensive: Yeah... I although I'm not sure how. I do remember that there was some new API for that.

**Comment by bilelmoussaoui** [comment]
> For the icon, I would just leave it out. Icons are mostly part of the .desktop file. I've never seen the icon tag used in the metainfo.
> 
> > By the way, is there a way to allow a flatpak access to dconf settings?
> 
> pensive Yeah... I although I'm not sure how. I do remember that there was some new API for that.

https://docs.flatpak.org/en/latest/sandbox-permissions.html#dconf-access

**Comment by JerryCasiano** [comment]
I'm guessing we would need

```
--filesystem=xdg-run/dconf
--filesystem=~/.config/dconf:rw
--talk-name=ca.desrt.dconf
--env=DCONF_USER_CONFIG_DIR=.config/dconf
```

Since we need to write to certain keys based on user input in the desktop preference pane.

**Comment by Eonfge** [comment]
> I'm guessing we would need
> 
> ```
> --filesystem=xdg-run/dconf
> --filesystem=~/.config/dconf:rw
> --talk-name=ca.desrt.dconf
> --env=DCONF_USER_CONFIG_DIR=.config/dconf
> ```
> 
> Since we need to write to certain keys based on user input in the desktop preference pane.

Not working for me, here is the experimental branch:
https://github.com/Eonfge/flathub/tree/org.gnome.FontManager_experimental

But according to the docs around DBus and DConf (these permissions should be enough to give FontManager access)

**Comment by JerryCasiano** [comment]
> Caveat Emptor: I know very little about DBus and DConf

Ha. Join the club... it's huge.

Anyway, noticed that the talk name for file roller is not right, so that's probably why that doesn't work.

Try it with org.gnome.ArchiveManager1

Though, I think that it's going to run into problems with access to temporary folders but I could be wrong. If it works you should be able to drag an archive onto the font list to install the fonts inside of it or by selecting it in the file chooser after hitting the add button.

As far as accessing / modifying gsettings on the host I'm not seeing anything promising on that front. I don't think Flatpak is ever going to be compatible with that use case.

**Comment by bilelmoussaoui** [comment]
There are a bunch of apps that do that on Flathub without any issues. I will have a look

**Comment by Eonfge** [comment]
> Try it with org.gnome.ArchiveManager1

Mmh. That seems to work partially... how you knew that it was called `ArchiveManager1` ? I used D-Feet to check the namespace, and it gave me `FileRoller`. Not that I can get the Unzipper to work properly

**Comment by JerryCasiano** [comment]
d-feet should show you, you just need to expand it on the right until you see the interface that has the actual functions exported by the application.

**Comment by JerryCasiano** [comment]
@bilelmoussaoui 

I would appreciate it, thank you.

I looked to see how something like Gnome Tweaks did it but couldn't find a Flatpak for it.

**Comment by Eonfge** [comment]
> d-feet should show you, you just need to expand it on the right until you see the interface that has the actual functions exported by the application.

Guess I was to fast to judge, but I must say that even with D-Feet, it's not easy to figure out

> I looked to see how something like Gnome Tweaks did it but couldn't find a Flatpak for it.

Yes, GNOME Tweaks is one of those applications which is so deeply connected to the main desktop experience, that I don't know if it even makes sense to package it as a Flatpak. For the same reason that you'll like not put GNOME Terminal or GNOME Files in a Flatpak... to much system integration.

Reminds me a bit of Hardware Info. Nice tool... but putting it in a Flatpak is almost self-defeating;
https://github.com/flathub/org.linux_hardware.hw-probe/blob/master/org.linux_hardware.hw-probe.yaml

**Comment by bilelmoussaoui** [comment]
You can't make apps that are not built with portals in mind to just work. There's Builder/KGX that have/are a nice terminal that just works as a flatpak. It's just requires more work as a lot of portals are still missing currently.

**Comment by Eonfge** [comment]
I've been busy for some days, so I didn't keep to much track on the current submission process. 

@JerryCasiano Do you feel that this is something to wait for, or do you think that we have a good first release here? It will take some work from your side to integrate DBus and FileRoller, but it might not be worth delaying that until that's done. My experience has been mixed with apps on Flathub. Some apps are surprisingly popular, like Asunder or Flacon, and others are not. Only after an initial release, will you know if Flatpak is a worthwhile time investment. 

Especially if you must rebuild some stuff, I think it is better to release a 90% functional package and wait to see what comes up later.

**Comment by JerryCasiano** [comment]
The file-roller integration will probably never work with Flatpak, so not really worth waiting on it. Eventually I'll just switch to using libarchive or gnome-autoar directly and that should solve that issue.

As far as the settings pane, I think that's another thing that simply isn't compatible with sandboxed applications.

In the next release, I'll probably detect that we're running from a Flatpak and put a note about it in the settings pane and also not log the file-roller message.

But for now, I think you just release it as is.

**Comment by JerryCasiano** [comment]
Closing this out as the Flatpak package has been published.

Thanks again!
